### PR TITLE
Use u32assert2 in u64 stdlib

### PR DESCRIPTION
**Pull request by tohrnii**
Addresses #217.
Builds on the work done by @DeimosTheMisthios in PR #252.
Modifies the following u64 stdlib procedures to use u32assert.2 instruction:
`u32assert4`,  `checked_add`, `checked_sub`, `checked_mul`, `checked_lt`, `checked_gt`, `checked_eqz`, `unchecked_div`, `checked_div`, `unchecked_mod`, `checked_mod`, `unchecked_divmod` and `checked_divmod`.

### CODE REVIEW

File: stdlib/asm/math/u64.masm [hunk 1]
```diff
@@ -156,8 +161,26 @@ end
...
+    movup.5
+    u32assert.2
+    u32overflowing_madd
+    movup.3
+    movup.2
+    u32overflowing_add
+    movup.2
+    add
     u32checked_or
     eq.0
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:185]: `u32checked_or` is relatively expensive. I think we should be able to replace it with a regular `or` (but double check me on this).

Also, my understanding is that the last few instructions are just checking if the top 3 values on the stack are zeros. And I think we know that all of them are smaller than $2^32$ - so, maybe we can do something a bit more efficient?

For example, could we replace:
```
movup.2
add
u32checked_or
eq.0
assert
```
with
```
add
add
eq.0
assert
```

File: stdlib/asm/math/u64.masm [hunk 2]
```diff
@@ -387,9 +431,8 @@ export.unchecked_div
     adv.u64div          # inject the quotient and the remainder into the advice tape
 
     push.adv.1          # read the quotient from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs
-    push.adv.1          # TODO: this can be optimized once we have u32assert2 instruction
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:434]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 3]
```diff
@@ -411,9 +454,8 @@ export.unchecked_div
     assert
 
     push.adv.1          # read the remainder from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs
-    push.adv.1
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:457]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 4]
```diff
@@ -456,9 +548,8 @@ export.unchecked_mod
     adv.u64div          # inject the quotient and the remainder into the advice tape
 
     push.adv.1          # read the quotient from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs
-    push.adv.1          # TODO: this can be optimized once we have u32assert2 instruction
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:551]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 5]
```diff
@@ -511,8 +601,58 @@ end
 # Stack transition looks as follows:
 # [b_hi, b_lo, a_hi, a_lo, ...] -> [c_hi, c_lo, ...], where c = a % b
 export.checked_mod
-    exec.u32assert4
-    exec.unchecked_mod
+    u32assert.2         # make sure divisor consists of 32-bit limbs
+    adv.u64div          # inject the quotient and the remainder into the advice tape
+
+    push.adv.1          # read the quotient from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:608]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 6]
```diff
@@ -580,8 +718,58 @@ end
 # Stack transition looks as follows:
 # [b_hi, b_lo, a_hi, a_lo, ...] -> [r_hi, r_lo, q_hi, q_lo ...], where r = a % b, q = a / b
 export.checked_divmod
-    exec.u32assert4
-    exec.unchecked_divmod
+    u32assert.2         # make sure divisor consists of 32-bit limbs
+    adv.u64div          # inject the quotient and the remainder into the advice tape
+
+    push.adv.1          # read the quotient from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:725]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 7]
```diff
@@ -549,9 +688,8 @@ export.unchecked_divmod
     assert
 
     push.adv.1          # read the remainder from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs 
-    push.adv.1
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:691]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 8]
```diff
@@ -525,9 +665,8 @@ export.unchecked_divmod
     adv.u64div          # inject the quotient and the remainder into the advice tape
 
     push.adv.1          # read the quotient from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs
-    push.adv.1          # TODO: this can be optimized once we have u32assert2 instruction
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:668]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 9]
```diff
@@ -442,8 +484,58 @@ end
 # Stack transition looks as follows:
 # [b_hi, b_lo, a_hi, a_lo, ...] -> [c_hi, c_lo, ...], where c = a // b
 export.checked_div
-    exec.u32assert4
-    exec.unchecked_div
+    u32assert.2         # make sure divisor is made of 32-bit limbs
+    adv.u64div          # inject the quotient and the remainder into the advice tape
+
+    push.adv.1          # read the quotient from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:491]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 10]
```diff
@@ -442,8 +484,58 @@ end
...
+    eq.0
+    assert
+    dup.4
+    dup.3
+    mul
+    eq.0
+    assert
+
+    push.adv.1          # read the remainder from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:514]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 11]
```diff
@@ -442,8 +484,58 @@ end
...
+    swap                # add remainder to the previous result; this also consumes the remainder
+    movup.3
+    u32overflowing_add
+    movup.3
+    movup.3
+    u32overflowing_add3
+    eq.0
+    assert
+
+    movup.4             # make sure the result we got is equal to the dividend
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:533]: I'd also add a comment here to explain that `u32assert.2`s are used to ensure that the dividend consists of two 32-bit limbs (at least that's what I think is happening).

File: stdlib/asm/math/u64.masm [hunk 12]
```diff
@@ -480,9 +571,8 @@ export.unchecked_mod
     assert
 
     push.adv.1          # read the remainder from the advice tape and make sure it consists of
-    u32assert           # 32-bit limbs
-    push.adv.1
-    u32assert
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:574]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 13]
```diff
@@ -511,8 +601,58 @@ end
...
+    dup.1               # add remainder to the previous result
+    movup.4
+    u32overflowing_add
+    movup.4
+    dup.3
+    u32overflowing_add3
+    eq.0
+    assert
+
+    movup.4             # make sure the result we got is equal to the dividend
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:650]: I'd also add a comment here to explain that `u32assert.2`s are used to ensure that the dividend consists of two 32-bit limbs (at least that's what I think is happening).

File: stdlib/asm/math/u64.masm [hunk 14]
```diff
@@ -511,8 +601,58 @@ end
...
+    eq.0
+    assert
+    dup.3
+    movup.3
+    mul
+    eq.0
+    assert
+
+    push.adv.1          # read the remainder from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:631]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 15]
```diff
@@ -580,8 +718,58 @@ end
...
+    eq.0
+    assert
+    dup.4
+    dup.3
+    mul
+    eq.0
+    assert
+
+    push.adv.1          # read the remainder from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:748]: You should be able to replace `push.adv.1 push.adv.1` with `push.adv.2`.

File: stdlib/asm/math/u64.masm [hunk 16]
```diff
@@ -580,8 +718,58 @@ end
...
+    dup.3
+    u32overflowing_add3
+    eq.0
+    assert
+
+    movup.6             # make sure the result we got is equal to the dividend
+    u32assert.2
+    assert_eq
+    movup.5
+    u32assert.2
```
Comment (bobbinth) [inline stdlib/asm/math/u64.masm:771]: I'd also add a comment here to explain that `u32assert.2`s are used to ensure that the dividend consists of two 32-bit limbs (at least that's what I think is happening).

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few small comments inline.

But also, I'm wondering whether we should just implement `checked_div` as:
```
exec.u32assert4
exec.unchecked_div
```
I think the difference between the current implementation and the above is just 3 cycles (out of 60+ cycles). And I'm wondering if 5% improvement is worth so much redundant code. I am inclined to say that it is not.

Same applies to `checked_mod` and `checked_divmod`.

File: stdlib/asm/math/u64.masm (hunk 1 above)
Comment (tohrnii) [reply stdlib/asm/math/u64.masm]: Thanks. I think it makes sense. Changed the sequence of instructions with your snippet.

File: stdlib/asm/math/u64.masm (hunk 2 above)
Comment (tohrnii) [reply stdlib/asm/math/u64.masm]: Thanks. Changed here and the rest of the places you suggested.

File: stdlib/asm/math/u64.masm (hunk 11 above)
Comment (tohrnii) [reply stdlib/asm/math/u64.masm]: Since this procedure has been changed now, this comment is no longer required.

**Comment by tohrnii** [comment]
Thank you @bobbinth for the review. Made suggested changes.
Re implementing `checked_div`, `checked_mod` and `checked_divmod` as `exec.u32assert4 exec.unchecked_div`, `exec.u32assert4 exec.unchecked_mod` and `exec.u32assert4 exec.unchecked_divmod`, I wasn't sure if the redundancy was worth it. I think it makes sense to keep it simple for now. We can re-evaluate if these procedure are used frequently in some cryptographic primitives in the future or if these procedures are frequently used when we have more visibility into user written miden assembly code. So I've changed these back to `exec.u32assert4 exec.unchecked_div` and similar for now.

**Pull request merged**